test/framework/src/main/java/org/elasticsearch/common/logging/TestThreadInfoPatternConverter.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.apache.logging.log4j.core.pattern.ConverterKeys;
+import org.apache.logging.log4j.core.pattern.LogEventPatternConverter;
+import org.apache.logging.log4j.core.pattern.PatternConverter;
+import org.apache.lucene.util.CloseableThreadLocal;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.test.ESIntegTestCase;
+
+/**
+ * Converts {@code %test_thread_info} in log4j patterns into information
+ * based on the loggin thread's name. If that thread is part of an
+ * {@link ESIntegTestCase} then this information is the node name.
+ */
+@Plugin(category = PatternConverter.CATEGORY, name = "TestInfoPatternConverter")
+@ConverterKeys({"test_thread_info"})
+public class TestThreadInfoPatternConverter extends LogEventPatternConverter {
+    /**
+     * Called by log4j2 to initialize this converter.
+     */
+    public static TestThreadInfoPatternConverter newInstance(final String[] options) {
+        if (options.length > 0) {
+            throw new IllegalArgumentException("no options supported but options provided: "
+                    + Arrays.toString(options));
+        }
+        return new TestThreadInfoPatternConverter();
+    }
+
+    private TestThreadInfoPatternConverter() {
+        super("TestInfo", "test_thread_info");
+    }
+
+    /**
+     * Cache of the string to output for this thread. It is a thread local
+     * because Elasticsearch's tests log synchronously so the formatting
+     * thread should be the same as the logging thread so the cache
+     * *should* only miss when a thread changes its name. Which they do.
```

Comment:
To add to this: only test threads change their name that I know of, but I figure it is worth checking all threads.